tsnake/grid.py

```diff
@@ -47,14 +62,19 @@ def __init__(self, point1: Point, point2: Point) -> None:
         super().__init__(point1, point2)
         self.intersections = list()
 
-    def add_intersection(self, point: Point) -> None:
+    def add_intersection(self, point: Point, element: Element) -> None:
         """
         Store the intersection in the edge
         Args:
         -------------------------------
         * point: Point(), point object representing where the intersection occured
+        * element: Element(), element that the grid intersected with
         -------------------------------
         """
+        # TODO: I have the snake element (and normal) as well as the intersection point
```

Comment:
Multiple snake elements can intersect a given grid edge. For a given snake element, when we find a grid element that it intersects:

First, compute the "sign" of this intersection. To do that, look at the lower number grid vertex of the grid edge (meaning, look at the grid vertex point that is smaller compared to the other one. If we sort points, one point is smaller. (3,3) is "smaller" than (3,4). You know what I mean...) We look at the lower numbered vertex because we want a common reference point to look at for all snake elements -> if they all intersect the same grid edge, they all look at the same grid vertex of that edge. Compare that point to the normal of the snake element. Is the grid vertex inside or outside of the snake? If it's inside the snake: give the intersection point a "+" sign. If it's outside the snake, give the intersection a "-" sign.

Next, after you have given this snake element a sign for the given intersection on the given grid edge, you want to check if there are any other intersections with this same grid edge.

Cases:
Case 1 - No other intersections. Then, just store this intersection in the list for the grid edge.
Case 2- Another intersection exists. Compare signs with the other intersection. If equal signs, this intersection point replaces the old intersection point. If opposite signs, remove both intersections and don't save anything.

This way, if you do this checking while going through the snake element by element, only one intersection is ever stored in the grid edge's list.

Final step: If the grid vertex inside the snake is "off", we store it in a queue for processing in Phase 2 (not sure if this has been done yet).